spv/sync.go

```diff
@@ -0,0 +1,1433 @@
+					fmatchidx = append(fmatchidx, i)
+					fmatchMu.Unlock()
+				}
+			}
+			wg.Done()
+		}
+		nworkers := 0
+		for i := idx; i < len(chain); i++ {
+			if fetched[i] != nil {
+				continue // Already have block
+			}
+			select {
+			case c <- i:
+			default:
+				if nworkers < runtime.NumCPU() {
+					nworkers++
+					wg.Add(1)
+					go worker()
+				}
+				c <- i
+			}
+		}
+		close(c)
+		wg.Wait()
+
+		if len(fmatches) != 0 {
+			blocks, err := rp.Blocks(ctx, fmatches)
+			if err != nil {
+				return nil, err
+			}
+			for j, b := range blocks {
+				i := fmatchidx[j]
+
+				// Perform context-free validation on the block.
+				// Disconnect peer when invalid.
+				err := validate.MerkleRoots(b)
+				if err != nil {
+					err = validate.DCP0005MerkleRoot(b)
+				}
+				if err != nil {
+					rp.Disconnect(err)
+					return nil, err
+				}
+				err = validate.RegularCFilter(b, chain[i].Filter)
+				if err != nil {
+					rp.Disconnect(err)
+					return nil, err
+				}
+
+				fetched[i] = b
+			}
+		}
+
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+
+		for i := idx; i < len(chain); i++ {
+			b := fetched[i]
+			if b == nil {
+				continue
+			}
+			matches, fadded := s.rescanBlock(b, walletID)
+			found[*chain[i].Hash] = matches
+			if len(fadded) != 0 {
+				idx = i + 1
+				filterData = fadded
+				continue FilterLoop
+			}
+		}
+		return found, nil
+	}
+	return found, nil
+}
+
+// handleBlockAnnouncements handles blocks announced through block invs or
+// headers messages by rp.  bmap should contain the full blocks of any
+// inventoried blocks, but may be nil in case the blocks were announced through
+// headers.
+func (s *Syncer) handleBlockAnnouncements(ctx context.Context, rp *p2p.RemotePeer, headers []*wire.BlockHeader,
+	bmap map[chainhash.Hash]*wire.MsgBlock) (err error) {
+
+	const opf = "spv.handleBlockAnnouncements(%v)"
+	defer func() {
+		if err != nil && ctx.Err() == nil {
+			op := errors.Opf(opf, rp.RemoteAddr())
+			err = errors.E(op, err)
+		}
+	}()
+
+	if len(headers) == 0 {
+		return nil
+	}
+
+	blockHashes := make([]*chainhash.Hash, 0, len(headers))
+	for _, h := range headers {
+		hash := h.BlockHash()
+		blockHashes = append(blockHashes, &hash)
+	}
+	filters, err := rp.CFilters(ctx, blockHashes)
+	if err != nil {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		return err
+	}
+
+	for key, w := range s.wallets {
+		newBlocks := make([]*wallet.BlockNode, 0, len(headers))
+		var bestChain []*wallet.BlockNode
+		var matchingTxs map[chainhash.Hash][]*wire.MsgTx
+		err = func() error {
+			defer s.sidechainMu.Unlock()
+			s.sidechainMu.Lock()
+
+			for i := range headers {
+				haveBlock, _, err := w.BlockInMainChain(ctx, blockHashes[i])
+				if err != nil {
+					return err
+				}
+				if haveBlock {
+					continue
+				}
+				n := wallet.NewBlockNode(headers[i], blockHashes[i], filters[i])
+				if s.sidechains.AddBlockNode(n) {
+					newBlocks = append(newBlocks, n)
```

Comment:
As explained in a previous review, `n` may not get added to the sidechain if it was already previously added by a previous wallet in this wallets loop. If it was added to the sidechain by a previous wallet and then it later got added to the wallet's mainchain, then it'd be pruned from the sidechain. Otherwise, it will remain on the sidechain.

I believe it's safer to use the approach introduced in https://github.com/raedahgroup/dcrlibwallet/pull/57/commits/00f2f95a736be8f64f53a8ba62425da0a6d678e3, specifically relating to using the `s.allWalletsHaveBlock` method.